### DISCUSSION TITLE
About this project

### DISCUSSION THREAD

**Discussion by mattgperry** [motiondivision/motionone#1] [Announcements]
Hey all!

First of all, massive thanks for becoming a sponsor. I'm trying something new here compared to my usual open source approach and I'm hoping to hit on a more sustainable development model. So thanks for being a part of that and showing your support!

So what is this project?

As you may know, I initially announced [motion.dev](https://motion.dev) as a learning resource. In the planning stages I identified a bunch of potential courses. Some of these would seem especially obvious, perhaps expected, like those revolving around JavaScript-based animation and specifically Framer Motion.

However, this was meant to be a fun side-project, so I wanted to concentrate on aspects of motion I was less familiar with, and take another look at technologies I had previously dismissed. This lead me to the native browser technologies, and part of that was [WAAPI](https://developer.mozilla.org/en-US/docs/Web/API/Web_Animations_API).

## WAAPI

WAAPI is a strange API. Intuitively, I would expect an API like

```javascript
animate(element, { opacity: 0 })
```

To animate `element.style.opacity` from its current `opacity` to `0`. To test what really happens, inspect any element on this page and type in the console `$0.animate({ opacity: 0 }, { duration: 300 })`

Spoiler: The animate fades out, and then `opacity` instantly snaps to its initial value (probably `1`). 

WAAPI has a concept of "fill". Which I guess is how an animation "fills out" values before and after it. So setting `fill` to `"forwards"` will leave the element with an expected `opacity` of `0`.

```javascript
$0.animate({ opacity: 0 }, { duration: 300, fill: "forwards" })
```

However, this is actively [discouraged by the specs](https://www.w3.org/TR/web-animations-1/#fill-behavior). Fill mode actually keeps the animation actively running and in memory indefinitely. You can verify this by writing `$0.style.opacity = 1`; no effect. 

To fix, it's recommended to set the final state of the animation when the animation finishes:

```javascript
element.animate({ transform: 'translateY(100px)' }, 200).finished(() => {
  element.style.transform = 'translateY(100px)';
});
```

Discovering this was the initial spark that made me feel time would be better spent fixing this behaviour with a wrapper library.

## Benefits of WAAPI

WAAPI comes with a few quite significant benefits that are quite hard to ignore.

### Performance

WAAPI runs animations of most values off the main thread. I believe it might if a value triggers layout (`width`, `top` etc) but we should only be animating transforms, `filter` and `opacity` for performance reasons anyway.

This means if lots of work is happening on the main thread, animations will remain smooth. 

As always, things aren't this simple. This famous [Greensock stress test](https://greensock.com/js/speed.html) shows that main thread animations *can* significantly out perform off-thread animations, but as most UIs aren't usually animating thousands of stars I suspect avoiding jank is the preferable trade-off.

### Interpolation

For a JavaScript library, animating between dimension value types like `"20px"` and `"calc(var(--width))"` is flaky and costly, both in terms of runtime performance and bundlesize. Converting between other value types like RGBA <-> HSLA is more straightforward but still negatively affects bundlesize.

In contrast, WAAPI has fully specs-compliant interpolators. So you can do all of the above at no extra cost. What's more, to animate between those dimension value types you'd need to set and measure the size of an element at the start of the animation. So `"calc(--var(width))"` would be converted to pixels once, but by the end of the animation `var(--width)` might have changed. WAAPI would account for changes every frame.

### File size

You get all of this and more for no extra filesize. To offer a fairly well-featured animation library with some quality of life improvements would be 0.5kbish.

## Drawbacks of WAAPI

However, nothing in life is free. WAAPI has a number of significant drawbacks. I think the trick is pressing against these as much as possible while finding a balance with the existing benefits.

### Individual transforms

The best-performing animation props are `opacity` and `transform`. But `transform` is really a collection of many props, so in all my animation libraries I've made it possible to animate them with shorthand syntax and animate them separately:

```jsx
<motion.div animate={{ x: 100, scaleX: 2 }} />
```

Until now this has been impossible with WAAPI. The spec binds itself tightly to the limitations of CSS and this doesn't allow for the individual setting of values.

But recently the [CSS Properties and Values API](https://developer.mozilla.org/en-US/docs/Web/API/CSS_Properties_and_Values_API/guide) offers us a way out. This would allow us to register individual transforms as CSS variables and animate those instead.

Animating these values individually would lose us the ability to animate them between unit types, but would gain us the ability to offer default unit types.

```javascript
// 👍 x could default to "px"
animate("div", [{ x: 10 }, { x: 100 }])

// 👎 Unit type interpolation on individual transforms
animate("div", [{ x: "10%" }, { x: 100 }])

// 👍 Different unit types with full transform string
animate("div", [{ transform: "translateX(10%)" }, { transform: "translateX(100px)" }])
```

This spec doesn't enjoy wide browser support just yet. But if we introduced an API restriction that individual transforms could be numbers-only we could provide a simple animation polyfill for an incredibly small bundlesize that would drive these CSS variables on the main thread in older browsers. 

### Custom easing and spring animations

WAAPI offers the easing functions it offers and nothing more. The most bizarre limitation of the upcoming AnimationWorklet is that this doesn't offer a way around this either, without some weird hacks.

However by driving `currentTime` we could again provide some small animation drivers that would scrub `currentTime`, trading in some main-thread performance for the powerful interpolation.

Alternatively, if individual transforms were guaranteed to be numbers we could precompute keyframes with animations not unlike Popmotion's spring and inertia animations, buying back that off-thread performance. This would also make velocity-based springs trivial as we could sample existing animations around `currentTime` if an animation is interuppted.

## Going forward

I've already created successful prototypes of the above features, so now I'm going to write a library around it. 

I want to start by offering a QoL wrapper for WAAPI that should run in at under 1kb. Then, to introduce individual transforms, before introducing spring animations. Hopefully for no more than 3-4kb.

 Once these basic features are in it could be fun to bring these capabilities to view libraries like React and Svelte. In the longer term it would be fun to look into timelines etc, but I'd like to keep it open-ended and in discussion with sponsors.
 
 Let me know what you think!

**Comment by souporserious** [comment]
Love all of this and appreciate the work you're doing here, Matt! Curious if you're thinking about how FLIP animations might play a role with the WAAPI?

Related to springs, some ideas we've tossed around at work were generating keyframes for spring animations that haven't been interrupted and then switching over to the main thread if it is interrupted. Not sure it's entirely worth that delicate balance or if it's even possible.

**Reply by mattgperry** [reply]
Hey man great to see you here 😃

What’s your reasons for switching to the main thread?

My current thinking is to limit spring animations to individual transforms as raw numbers (for bundle size). It also makes the most sense to preserve velocity as px/deg etc. Although we could support unit types later. 

Popmotion’s spring uses a velocity equation to figure out whether to stop on a given frame. I’ve modified that to expose this value - then if an animation is interrupted you can take the spring generator and scrub it to currentTime to get the velocity. 

I’m also going to look into using the simpler WebKit spring for a much smaller bundle size. If this one doesn’t calculate velocity the same way we can sample the animation at current time and current time - 16ms and grab the velocity from that. 

In all these approaches I was going to precalculate keyframes too though

**Reply by mattgperry** [reply]
Oh yeah and as for FLIP, I probably won’t do anything specific for a while. Even in this first release it’s capable of doing the classic FLIP process. I might add some FLIP if I do a React library as there’s value in simplifying the API. But I’m still fairly convinced Framer Motion-style are one of the things WAAPI will never do in its current form

**Reply by souporserious** [reply]
Whoa, scrubbing it to currentTime makes more sense now, yeah so no need to ever jump to the main thread 🤯 👏

**Comment by FallDownTheSystem** [comment]
Fantastic library! I'm excited for the declarative binding for the view frameworks, especially for Vue since it lacks good animation libraries like Framer Motion.

Now that a lot of the initial plans have been implemented, I was wondering if something akin to FLIP might be on the horizon?

**Comment by stepanjakl** [comment]
Hi @mattgperry, I would also be curious to know if there are any plans to incorporate FLIP transitions to the Motion One library?

**Reply by mattgperry** [reply]
No it's quite unlikely. It'd be possible to do some basic FLIP stuff with Motion One but since doing proper layout animations in Framer Motion with scale correction etc I wouldn't want to spend time doing an inferior version of that. 

It might be as the Shared Element Transitions API matures I might write something based on that, to make that experience nicer.

**Reply by stepanjakl** [reply]
Got it, thank you for the answer.
